### ISSUE TITLE
"Unexpected Frame" send during parallel benchmark

### ISSUE THREAD

**Issue by devigned** [vcabbage/amqp#6]
I'm running a parallel benchmark sending to a queue similar to the following:
```go
session, err := sb.client.NewSession() // client is *amqp.Client -- the client connection is reused
if err != nil {
	return err
}
defer session.Close()

sender, err := session.NewSender(amqp.LinkAddress(entityPath))
if err != nil {
	return err
}
defer sender.Close()

return sender.Send(ctx, msg)
```

Received the following:
```
Unexpected frame: &{*Error(nil)}
Unexpected frame: *performClose{Error: *Error{Condition: amqp:session:handle-in-use, Description: The handle '0' is already associated with object 'session55149196'., Info: map[]}}
```
The source of the error is: https://github.com/vcabbage/amqp/blob/a6e33075b4f688b0d4019d6d48c09aa664ca0e05/client.go#L474

Any thoughts on why?

**Comment by devigned** [comment]
I'm wondering if this excerpt is the reason (via: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-protocol-guide)
> Azure Service Bus currently uses exactly one session for each connection. The Service Bus maximum frame-size is 262,144 bytes (256 K bytes) for Service Bus Standard and Event Hubs. It is 1,048,576 (1 MB) for Service Bus Premium. Service Bus does not impose any particular session-level throttling windows, but resets the window regularly as part of link-level flow control (see the next section).

I'm going to spin up a premium instance and see if it replicates.

**Comment by devigned** [comment]
Running the test against Premium Service Bus exhibits the same behavior.

**Comment by vcabbage** [comment]
I don't think I've tried opening multiple sessions on a link before and it seems the the channel number handling is wrong. The fix for the immediate issue is relatively simple, but I'd like to fix the maximum channel handling too. I hope to complete the fix tonight.

As an aside, is there advantage to using a single connection with multiple sessions vs multiple connections with a single session each? There are certainly tradeoffs with each approach, I'm curious if you're optimizing for anything in particular.

**Comment by devigned** [comment]
I have been trying different connection, session and link combinations to see what performs the best. I’ll take a look at the .Net implementation and see how it was done there.

**Comment by vcabbage** [comment]
I pushed https://github.com/vcabbage/amqp/tree/channels with a fix. Would you mind testing it before I merge?

You'll probably need to adjust the channel-max with the newly added `ConnMaxChannels()` option.

The current default channel-max of 1 is probably too conservative. I may change it to use the spec default of 65535, I can't think of a reason not to. Do you know if the .Net implementation has a default?

FYI, I'll likely be changing that option before I merge the branch (see my comment on #7 for details), but I want to think on it a bit more before doing so. Normally I'm pretty aggressive about rebasing my working branches, but I'll wait until you've had a chance to test it to avoid causing you any issues with pulling/updating it if it needs changes.

**Comment by devigned** [comment]
The change is behaving as expected with regard to channels and links. 

I'm seeing some oddness when running a parallel send benchmark. In the test output below, the first test runs the send benchmark serially like so:
```go
func BenchmarkSend(b *testing.B) {
        // setup
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		sb.Send(context.Background(), queueName, &amqp.Message{
			Data: []byte("Hello!"),
		})
	}
	b.StopTimer()
        // clean up
}
```
The second test runs the benchmark as follows:
```go
func BenchmarkSend(b *testing.B) {
        // setup
	b.ResetTimer()
	b.RunParallel(func(pb *testing.PB){
		for pb.Next() {
			err = sb.Send(context.Background(), queueName, &amqp.Message{
				Data: []byte("Hello!"),
			})
			if err != nil {
				b.Fail()
			}
		}
	})
	b.StopTimer()
        // clean up
}
```
I'm running the benchmarks against a premium service bus instance. Let me know if you want me to collect traffic or anything else that would assist in debugging.
```
david at warmachine in ~/go/src/github.com/Azure/azure-service-bus-go on simple-send-receive [!]
$ go test -bench ... -v -timeout 100s
=== RUN   TestServiceBusSuite
=== RUN   TestServiceBusSuite/TestQueue
--- PASS: TestServiceBusSuite (7.48s)
    --- PASS: TestServiceBusSuite/TestQueue (6.05s)
=== RUN   TestCreateFromConnectionString
--- PASS: TestCreateFromConnectionString (0.24s)
goos: darwin
goarch: amd64
pkg: github.com/Azure/azure-service-bus-go
BenchmarkSend-4   	    3000	    396945 ns/op
PASS
ok  	github.com/Azure/azure-service-bus-go	34.093s


david at warmachine in ~/go/src/github.com/Azure/azure-service-bus-go on simple-send-receive [!]
$ go test -bench ... -v -timeout 100s
=== RUN   TestServiceBusSuite
=== RUN   TestServiceBusSuite/TestQueue
--- PASS: TestServiceBusSuite (8.24s)
    --- PASS: TestServiceBusSuite/TestQueue (6.69s)
=== RUN   TestCreateFromConnectionString
--- PASS: TestCreateFromConnectionString (0.29s)
goos: darwin
goarch: amd64
pkg: github.com/Azure/azure-service-bus-go
BenchmarkSend-4   	SIGQUIT: quit
PC=0x105e27b m=0 sigcode=0

goroutine 0 [idle]:
runtime.mach_semaphore_wait(0x1003, 0x10222ed, 0x16fbe40, 0x0, 0xc4204e5800, 0x16fbb00, 0x7ffeefbfea10, 0x1058ae3, 0xffffffffffffffff, 0xc4204e5800, ...)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/sys_darwin_amd64.s:445 +0xb
runtime.semasleep1(0xffffffffffffffff, 0xc4204e5800)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/os_darwin.go:413 +0x52
runtime.semasleep.func1()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/os_darwin.go:432 +0x33
runtime.systemstack(0x7ffeefbfea38)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/asm_amd64.s:360 +0xab
runtime.semasleep(0xffffffffffffffff, 0xc420032b60)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/os_darwin.go:431 +0x44
runtime.notesleep(0x16fc110)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/lock_sema.go:167 +0xe9
runtime.stopm()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:1680 +0xe5
runtime.findrunnable(0xc42002e000, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:2135 +0x4d2
runtime.schedule()
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:2255 +0x12c
runtime.park_m(0xc420450000)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/proc.go:2318 +0xb6
runtime.mcall(0x7ffeefbfec20)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/asm_amd64.s:286 +0x5b

goroutine 1 [chan receive, 2 minutes]:
testing.(*B).doBench(0xc420001800, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:261 +0x8e
testing.(*benchContext).processBench(0xc4200eaca0, 0xc420001800)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:447 +0x229
testing.(*B).run(0xc420001800, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:251 +0x9d
testing.(*B).Run(0xc420075200, 0x1486958, 0xd, 0x149dc68, 0x10dfe00)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:514 +0x295
testing.runBenchmarks.func1(0xc420075200)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:417 +0x60
testing.(*B).runN(0xc420075200, 0x1)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:141 +0xb2
testing.runBenchmarks(0x1491265, 0x25, 0xc4200eac40, 0x16f04f0, 0x1, 0x1, 0xc420081050)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:423 +0x49e
testing.(*M).Run(0xc42046df18, 0xc42005df70)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/testing.go:927 +0x31c
main.main()
	github.com/Azure/azure-service-bus-go/_test/_testmain.go:48 +0xdb

goroutine 82 [semacquire, 2 minutes]:
sync.runtime_Semacquire(0xc4201ce36c)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc4201ce360)
	/usr/local/Cellar/go/1.9.2/libexec/src/sync/waitgroup.go:131 +0x72
testing.(*B).RunParallel(0xc420001800, 0xc42013c5c0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:628 +0x1b9
github.com/Azure/azure-service-bus-go.BenchmarkSend(0xc420001800)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus_test.go:130 +0x3fd
testing.(*B).runN(0xc420001800, 0x1f4)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:141 +0xb2
testing.(*B).launch(0xc420001800)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:291 +0x84
created by testing.(*B).doBench
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:260 +0x70

goroutine 74 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*conn).connWriter(0xc4200bab00)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:516 +0x1f0
created by github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.newConn
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:218 +0x437

goroutine 88 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*Sender).Send(0xc42022b960, 0x16c3d60, 0xc42001a0d8, 0xc4203330a0, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:276 +0x6e5
github.com/Azure/azure-service-bus-go.(*serviceBus).Send(0xc420082340, 0x16c3d60, 0xc42001a0d8, 0xc42014b240, 0x14, 0xc4203330a0, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus.go:302 +0x92
github.com/Azure/azure-service-bus-go.BenchmarkSend.func2(0xc4200eb060)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus_test.go:132 +0x1b6
testing.(*B).RunParallel.func1(0xc4201ce360, 0xc4201ce358, 0xc4201ce350, 0xc420001800, 0xc42013c5c0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:625 +0x9c
created by testing.(*B).RunParallel
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:618 +0x199

goroutine 72 [IO wait, 2 minutes]:
internal/poll.runtime_pollWait(0x1ab0d80, 0x72, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/runtime/netpoll.go:173 +0x57
internal/poll.(*pollDesc).wait(0xc420118f98, 0x72, 0xffffffffffffff00, 0x16c0a20, 0x16bc540)
	/usr/local/Cellar/go/1.9.2/libexec/src/internal/poll/fd_poll_runtime.go:85 +0xae
internal/poll.(*pollDesc).waitRead(0xc420118f98, 0xc420421000, 0x1000, 0x1000)
	/usr/local/Cellar/go/1.9.2/libexec/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc420118f80, 0xc420421000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/internal/poll/fd_unix.go:126 +0x18a
net.(*netFD).Read(0xc420118f80, 0xc420421000, 0x1000, 0x1000, 0x10a7592, 0x1449ba0, 0x3)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/fd_unix.go:202 +0x52
net.(*conn).Read(0xc420316100, 0xc420421000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/net/net.go:176 +0x6d
crypto/tls.(*block).readFromUntil(0xc420131560, 0x1a701a0, 0xc420316100, 0x5, 0xc420316100, 0xc42005fcf8)
	/usr/local/Cellar/go/1.9.2/libexec/src/crypto/tls/conn.go:488 +0x95
crypto/tls.(*Conn).readRecord(0xc42012b180, 0x149e617, 0xc42012b2a0, 0xbe90ae5c44790118)
	/usr/local/Cellar/go/1.9.2/libexec/src/crypto/tls/conn.go:590 +0xe0
crypto/tls.(*Conn).Read(0xc42012b180, 0xc42024ac00, 0x600, 0x600, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/crypto/tls/conn.go:1134 +0x110
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*frameReader).Read(0xc4201e8760, 0xc42024ac00, 0x600, 0x600, 0x5a6241b5, 0xc42005fd78, 0x1201531)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:363 +0x55
bytes.(*Buffer).ReadFrom(0xc420112f50, 0x16be4e0, 0xc4201e8760, 0x16fb680, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.2/libexec/src/bytes/buffer.go:209 +0x177
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*conn).connReader(0xc4200bab00)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:391 +0x86c
created by github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.newConn
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:202 +0x353

goroutine 73 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*conn).mux(0xc4200bab00)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:286 +0x544
created by github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.newConn
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/conn.go:217 +0x415

goroutine 90 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*link).mux(0xc4204300d0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:893 +0x270
created by github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.newLink
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:749 +0x8c2

goroutine 89 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*Session).mux(0xc42013c040, 0xc4202c61e0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:394 +0x838
created by github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*Client).NewSession
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:118 +0x50d

goroutine 87 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*Sender).Send(0xc42022b960, 0x16c3d60, 0xc42001a0d8, 0xc4200c2230, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:276 +0x6e5
github.com/Azure/azure-service-bus-go.(*serviceBus).Send(0xc420082340, 0x16c3d60, 0xc42001a0d8, 0xc42014b240, 0x14, 0xc4200c2230, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus.go:302 +0x92
github.com/Azure/azure-service-bus-go.BenchmarkSend.func2(0xc4201611e0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus_test.go:132 +0x1b6
testing.(*B).RunParallel.func1(0xc4201ce360, 0xc4201ce358, 0xc4201ce350, 0xc420001800, 0xc42013c5c0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:625 +0x9c
created by testing.(*B).RunParallel
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:618 +0x199

goroutine 86 [select, 2 minutes]:
github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp.(*Sender).Send(0xc42022b960, 0x16c3d60, 0xc42001a0d8, 0xc420113960, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/vendor/pack.ag/amqp/client.go:276 +0x6e5
github.com/Azure/azure-service-bus-go.(*serviceBus).Send(0xc420082340, 0x16c3d60, 0xc42001a0d8, 0xc42014b240, 0x14, 0xc420113960, 0x0, 0x0)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus.go:302 +0x92
github.com/Azure/azure-service-bus-go.BenchmarkSend.func2(0xc420204900)
	/Users/david/go/src/github.com/Azure/azure-service-bus-go/servicebus_test.go:132 +0x1b6
testing.(*B).RunParallel.func1(0xc4201ce360, 0xc4201ce358, 0xc4201ce350, 0xc420001800, 0xc42013c5c0)
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:625 +0x9c
created by testing.(*B).RunParallel
	/usr/local/Cellar/go/1.9.2/libexec/src/testing/benchmark.go:618 +0x199

rax    0xe
rbx    0x16fc000
rcx    0x7ffeefbfe9b0
rdx    0x7ffeefbfea38
rdi    0x1003
rsi    0xc4204e58000001
rbp    0x7ffeefbfe9e8
rsp    0x7ffeefbfe9b0
r8     0x6
r9     0x16fbb00
r10    0x2
r11    0x286
r12    0x1774152eff7c4
r13    0x0
r14    0x105bba0
r15    0x0
rip    0x105e27b
rflags 0x286
cs     0x7
fs     0x0
gs     0x0
*** Test killed with quit: ran too long (2m40s).
FAIL	github.com/Azure/azure-service-bus-go	160.026s
```

**Comment by vcabbage** [comment]
I haven't used `b.RunParallel()` before, but based on the name and the docs I assume the benchmark will attempt to concurrently send on the same link. `Sender.Send()` isn't safe for concurrent use. I wrote it that way because, as I remember, the spec doesn't allow interleaving transfer frames for different messages on the same link.

**Comment by devigned** [comment]
Good to know. I think you can disregard the concurrency issue then. It appears all is working as intended at this point. I'll close the issue.

**Comment by devigned** [comment]
I did a little digging into the .NET version of the Service Bus and AMQP libs. The queue messaging object hierarchy looks like [QueueClient](https://github.com/Azure/azure-service-bus-dotnet/blob/a51efe26a2b742a01443777e584cebed4ac97fdb/src/Microsoft.Azure.ServiceBus/QueueClient.cs) -> [MessageSender](https://github.com/Azure/azure-service-bus-dotnet/blob/a51efe26a2b742a01443777e584cebed4ac97fdb/src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs) -> [SendAndReceiveLinkCreator](https://github.com/Azure/azure-service-bus-dotnet/blob/a51efe26a2b742a01443777e584cebed4ac97fdb/src/Microsoft.Azure.ServiceBus/Amqp/AmqpSendReceiveLinkCreator.cs) --> ((Receiver Session, Receiver Link), (Sender Session, Sender Link)).

Basically, the progression is as such. It appears to have a Singleton connection, then one session per link. Link and Session lifespan are paired. When the link is closed, the corresponding session is also closed.

**Comment by vcabbage** [comment]
Interesting, it seems that most implementations expose a method to directly create a sender/receiver link from the connection/client, implicitly creating a session. Some also allow explicit session creation.

I can see value in hiding the details around sessions in higher level libraries.

As this library is intended to be low-level, I think the flexibility of explicit session creation is appropriate.

Helper methods for directly creating senders/receivers from `*Conn` are possible, but I'd like to keep the API as small as is reasonable. For now I'm going to defer adding helpers until a strong case is made for their inclusion.

**Comment by devigned** [comment]
Agreed. Explicit session creation from this library is essential for building higher level libraries upon it.

I wouldn't worry about creating the higher level constructs. That's what I'm going to focus on in the Service Bus and Event Hub libraries.

IMHO, a valuable focus, among others, for this library would be adherence to the specification and performance.
